File: src/main/java/eu/qwsome/sql/builder/ConditionBuilder.java
package eu.qwsome.sql.builder;


import eu.qwsome.sql.condition.Condition;

import java.util.Objects;

/**
 * This builder uses left associativity when appendeing conditions.
 *
 * @author Lukáš Kvídera
 */
public class ConditionBuilder {

  private Condition root;

  /**
   * Creates a new instance of {@link ConditionBuilder}
   *
   * @return {@link ConditionBuilder}
   */
  public static ConditionBuilder create() {
    return new ConditionBuilder();
  }

  /**
   * Appends another condition to current context with AND if applicable.
   *
   * @param another condition to be composed
   * @return this builder
   */
  public ConditionBuilder and(final Condition another) {
    Objects.requireNonNull(another, "provide another condition");
    this.root = this.root == null ? another : this.root.and(another);
    return this;
  }

  /**
   * Appends another condition to current context with OR if applicable.
   *
   * @param another condition to be composed
   * @return this builder
   */
  public ConditionBuilder or(final Condition another) {
    Objects.requireNonNull(another, "provide another condition");
    this.root = this.root == null ? another : this.root.or(another);
    return this;
  }

  /**
   * @return condition or null if empty
   */
  public Condition build() {
    return this.root;
  }
}

File: src/main/kotlin/eu/qwsome/sql/builder/ConditionBuilder.kt
/*
 * This builder uses left associativity when appending conditions.
 *
 * @author Lukáš Kvídera
 */
package eu.qwsome.sql.builder

import eu.qwsome.sql.condition.Condition

class ConditionBuilder private constructor() {

    private var root: Condition? = null

    /**
     * Appends another condition to current context with AND if applicable.
     *
     * @param another condition to be composed
     * @return this builder
     */
    fun and(another: Condition): ConditionBuilder = apply {
        root = root?.and(another) ?: another
    }

    /**
     * Appends another condition to current context with OR if applicable.
     *
     * @param another condition to be composed
     * @return this builder
     */
    fun or(another: Condition): ConditionBuilder = apply {
        root = root?.or(another) ?: another
    }

    /**
     * @return condition or null if empty
     */
    fun build(): Condition? = root

    companion object {
        /**
         * Creates a new instance of [ConditionBuilder]
         *
         * @return [ConditionBuilder]
         */
        @JvmStatic
        fun create(): ConditionBuilder = ConditionBuilder()
    }
}
